In [1]:
import sklearn
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
pd.options.display.max_columns = 110


In [45]:
byCategory=pd.read_csv('0614byCategory.csv')

/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3049: DtypeWarning: Columns (25) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [46]:
##將檔期時間調整為至少一個月(檔期開始時間最晚為2019-03-30)
byCategory=byCategory[byCategory.business_hour_order_time_s<'2019-03-30']

In [47]:
##找出檔期長度超過3年,檔期長度即設為1100天
byCategory.loc[byCategory.campagin_period>1100,'campagin_period']=1100

In [48]:
byCategory=byCategory[['business_hour_guid','product_launch_month', 'product_launch_hour',
       'product_launch_weekday','code', 'parent_code', 'is_Presales', 'is_Longtern',  
        'enable_isp','is_Discount','installment_3months', 'installment_6months', 'installment_12months',
       'deny_installment', 
       'item_orig_price', 'item_price','item_default_daily_amount',
       'quantity_multiplier','shippingdate', 'freight_amount',
        'campagin_period', 'order_deliver_days','unit_price',
       'category_unit_price', 'compare_cate_unit_price',
       'guid_subtotal', 'quantity_sold_guid', 'subtotal_allguid',
       'quantity_sold_allguid', 'avg_unit_price', 'price_gap',
        'hottimes']]


In [49]:
byCategory['unit_price']=byCategory['item_price']/byCategory['quantity_multiplier']


In [50]:
byCategory.columns

Index(['business_hour_guid', 'product_launch_month', 'product_launch_hour',
       'product_launch_weekday', 'code', 'parent_code', 'is_Presales',
       'is_Longtern', 'enable_isp', 'is_Discount', 'installment_3months',
       'installment_6months', 'installment_12months', 'deny_installment',
       'item_orig_price', 'item_price', 'item_default_daily_amount',
       'quantity_multiplier', 'shippingdate', 'freight_amount',
       'campagin_period', 'order_deliver_days', 'unit_price',
       'category_unit_price', 'compare_cate_unit_price', 'guid_subtotal',
       'quantity_sold_guid', 'subtotal_allguid', 'quantity_sold_allguid',
       'avg_unit_price', 'price_gap', 'hottimes'],
      dtype='object')

In [51]:
df=byCategory
df.shape

(224074, 32)

In [55]:
#發現 code,category_unit_price,compare_cate_unit_price 有nan值 因為分類關係
df.code=df.code.fillna(0)
df.shippingdate=df.shippingdate.fillna(0)
df.isna().sum()

business_hour_guid           0
product_launch_month         0
product_launch_hour          0
product_launch_weekday       0
code                         0
parent_code                  0
is_Presales                  0
is_Longtern                  0
enable_isp                   0
is_Discount                  0
installment_3months          0
installment_6months          0
installment_12months         0
deny_installment             0
item_orig_price              0
item_price                   0
item_default_daily_amount    0
quantity_multiplier          0
shippingdate                 0
freight_amount               0
campagin_period              0
order_deliver_days           0
unit_price                   0
category_unit_price          0
compare_cate_unit_price      0
guid_subtotal                0
quantity_sold_guid           0
subtotal_allguid             0
quantity_sold_allguid        0
avg_unit_price               0
price_gap                    0
hottimes                     0
dtype: i

In [56]:
df['category_unit_price']=df['category_unit_price'].fillna(df['unit_price'])

In [57]:
df['compare_cate_unit_price']=df['compare_cate_unit_price'].fillna(1)

## 將檔案順序打散

In [58]:
df=df.sample(frac=1, replace=False, random_state=1)

In [59]:
df=df.reset_index(drop=True)

將資料分為輸入變數與目標變數

## 特徵欄位進行標準化

In [60]:
df.head(3)

business_hour_guid  product_launch_month  \
0  AB1BA373-E211-4206-A9F1-4A11F4DB5314                     7   
1  F0AB8354-EB2A-44F6-87D7-F71DAFFF1D28                     6   
2  D7EE78D5-DE57-4486-881C-1DF8D8B220BC                     9   

   product_launch_hour  product_launch_weekday    code  parent_code  \
0                    7                       3   536.0        536.0   
1                    7                       3   166.0        166.0   
2                    7                       2  1604.0        166.0   

   is_Presales  is_Longtern  enable_isp  is_Discount  installment_3months  \
0            0            0           0            1                    1   
1            0            0           0            1                    0   
2            0            0           0            1                    1   

   installment_6months  installment_12months  deny_installment  \
0                    1                     1                 0   
1                    0                     0                 0   
2                    0                     0                 0   

   item_orig_price  item_price  item_default_daily_amount  \
0           3000.0      1088.0                         20   
1            890.0       239.0                         20   
2            799.0       429.0                         20   

   quantity_multiplier  shippingdate  freight_amount  campagin_period  \
0                 10.0           0.0             0.0               21   
1                  1.0           0.0             0.0               13   
2                  1.0           7.0             0.0               21   

   order_deliver_days  unit_price  category_unit_price  \
0                   0       108.8                299.0   
1                   0       239.0                359.0   
2                   0       429.0                319.0   

   compare_cate_unit_price  guid_subtotal  quantity_sold_guid  \
0                    0.364        16320.0               150.0   
1                    0.666         4302.0                18.0   
2                    1.345         2574.0                 6.0   

   subtotal_allguid  quantity_sold_allguid  avg_unit_price  price_gap  \
0               0.0                    0.0           108.8      191.2   
1               0.0                    0.0           239.0      651.0   
2               0.0                    0.0           429.0      370.0   

   hottimes  
0         0  
1         1  
2         0

In [61]:
target=pd.read_csv('target.csv')
code=pd.read_csv('code_time.csv')
target.head(3)

business_hour_guid  total_sales_amount  sales_7days  \
0  00000793-A001-4C8C-9196-3E6971181246               399.0        399.0   
1  00001911-C3DF-4143-BC25-8AA5FA1A9AF4              1150.0       1150.0   
2  00001928-AE75-4AA9-8A47-4BDE96D08C2D             23040.0          0.0   

   ratio_7days  sales_14days  ratio_14days  sales_30days  ratio_30days  \
0          1.0         399.0           1.0         399.0           1.0   
1          1.0        1150.0           1.0        1150.0           1.0   
2          0.0       11520.0           0.5       23040.0           1.0   

   sales_60days  ratio_60days  sales_180days  ratio_180days  
0         399.0           1.0          399.0            1.0  
1        1150.0           1.0         1150.0            1.0  
2       23040.0           1.0        23040.0            1.0

In [62]:
df=pd.merge(df,target,how='left')
df.head(3)

business_hour_guid  product_launch_month  \
0  AB1BA373-E211-4206-A9F1-4A11F4DB5314                     7   
1  F0AB8354-EB2A-44F6-87D7-F71DAFFF1D28                     6   
2  D7EE78D5-DE57-4486-881C-1DF8D8B220BC                     9   

   product_launch_hour  product_launch_weekday    code  parent_code  \
0                    7                       3   536.0        536.0   
1                    7                       3   166.0        166.0   
2                    7                       2  1604.0        166.0   

   is_Presales  is_Longtern  enable_isp  is_Discount  installment_3months  \
0            0            0           0            1                    1   
1            0            0           0            1                    0   
2            0            0           0            1                    1   

   installment_6months  installment_12months  deny_installment  \
0                    1                     1                 0   
1                    0                     0                 0   
2                    0                     0                 0   

   item_orig_price  item_price  item_default_daily_amount  \
0           3000.0      1088.0                         20   
1            890.0       239.0                         20   
2            799.0       429.0                         20   

   quantity_multiplier  shippingdate  freight_amount  campagin_period  \
0                 10.0           0.0             0.0               21   
1                  1.0           0.0             0.0               13   
2                  1.0           7.0             0.0               21   

   order_deliver_days  unit_price  category_unit_price  \
0                   0       108.8                299.0   
1                   0       239.0                359.0   
2                   0       429.0                319.0   

   compare_cate_unit_price  guid_subtotal  quantity_sold_guid  \
0                    0.364        16320.0               150.0   
1                    0.666         4302.0                18.0   
2                    1.345         2574.0                 6.0   

   subtotal_allguid  quantity_sold_allguid  avg_unit_price  price_gap  \
0               0.0                    0.0           108.8      191.2   
1               0.0                    0.0           239.0      651.0   
2               0.0                    0.0           429.0      370.0   

   hottimes  total_sales_amount  sales_7days  ratio_7days  sales_14days  \
0         0             16320.0       3264.0        0.200       11968.0   
1         1              4302.0       2390.0        0.556        4302.0   
2         0              2574.0       1287.0        0.500        1287.0   

   ratio_14days  sales_30days  ratio_30days  sales_60days  ratio_60days  \
0         0.733       16320.0           1.0       16320.0           1.0   
1         1.000        4302.0           1.0        4302.0           1.0   
2         0.500        2574.0           1.0        2574.0           1.0   

   sales_180days  ratio_180days  
0        16320.0            1.0  
1         4302.0            1.0  
2         2574.0            1.0

In [63]:
df=pd.merge(df,code)
df.head(3)

business_hour_guid  product_launch_month  \
0  AB1BA373-E211-4206-A9F1-4A11F4DB5314                     7   
1  9584B6E4-264C-4FFC-8EF6-5223301E692A                     9   
2  124C1F67-488B-4B55-B449-C45650BDECC7                     4   

   product_launch_hour  product_launch_weekday   code  parent_code  \
0                    7                       3  536.0        536.0   
1                    7                       4  536.0        536.0   
2                    7                       1  536.0        536.0   

   is_Presales  is_Longtern  enable_isp  is_Discount  installment_3months  \
0            0            0           0            1                    1   
1            0            1           0            1                    1   
2            0            0           0            1                    1   

   installment_6months  installment_12months  deny_installment  \
0                    1                     1                 0   
1                    0                     0                 0   
2                    1                     1                 0   

   item_orig_price  item_price  item_default_daily_amount  \
0           3000.0      1088.0                         20   
1           1380.0       799.0                         20   
2           1718.0      1150.0                         20   

   quantity_multiplier  shippingdate  freight_amount  campagin_period  \
0                 10.0           0.0             0.0               21   
1                  1.0           0.0             0.0               81   
2                  1.0           0.0             0.0               21   

   order_deliver_days  unit_price  category_unit_price  \
0                   0       108.8                299.0   
1                   0       799.0                299.0   
2                   0      1150.0                299.0   

   compare_cate_unit_price  guid_subtotal  quantity_sold_guid  \
0                    0.364        16320.0               150.0   
1                    2.672         3995.0                 5.0   
2                    3.846         1150.0                 1.0   

   subtotal_allguid  quantity_sold_allguid  avg_unit_price  price_gap  \
0               0.0                    0.0           108.8      191.2   
1               0.0                    0.0           799.0      581.0   
2               0.0                    0.0          1150.0      568.0   

   hottimes  total_sales_amount  sales_7days  ratio_7days  sales_14days  \
0         0             16320.0       3264.0          0.2       11968.0   
1         0              3995.0        799.0          0.2         799.0   
2         0              1150.0       1150.0          1.0        1150.0   

   ratio_14days  sales_30days  ratio_30days  sales_60days  ratio_60days  \
0         0.733       16320.0           1.0       16320.0           1.0   
1         0.200         799.0           0.2        3196.0           0.8   
2         1.000        1150.0           1.0        1150.0           1.0   

   sales_180days  ratio_180days  code_total_sales_amount  code_sales_7days  \
0        16320.0            1.0              192491020.0        67838255.0   
1         3995.0            1.0              192491020.0        67838255.0   
2         1150.0            1.0              192491020.0        67838255.0   

   code_ratio_7days  code_sales_14days  code_ratio_14days  code_sales_30days  \
0             0.352        101498632.0              0.527        140193016.0   
1             0.352        101498632.0              0.527        140193016.0   
2             0.352        101498632.0              0.527        140193016.0   

   code_ratio_30days  code_sales_60days  code_ratio_60days  \
0              0.728        159588691.0              0.829   
1              0.728        159588691.0              0.829   
2              0.728        159588691.0              0.829   

   code_sales_180days  code_ratio_180days  code_max_7days  code_min_7days  \
0         18214011

In [64]:
parent_code=code
parent_code.columns=['parent_code', 'parent_code_total_sales_amount', 'parent_code_sales_7days', 'parent_code_ratio_7days',
       'parent_code_sales_14days', 'parent_code_ratio_14days', 'parent_code_sales_30days',
       'parent_code_ratio_30days', 'parent_code_sales_60days', 'parent_code_ratio_60days',
       'parent_code_sales_180days', 'parent_code_ratio_180days', 'parent_code_max_7days',
       'parent_code_min_7days', 'parent_code_median_7days', 'parent_code_mean_7days',
       'parent_code_max_14days', 'parent_code_min_14days', 'parent_code_median_14days',
       'parent_code_mean_14days', 'parent_code_max_30days', 'parent_code_min_30days',
       'parent_code_median_30days', 'parent_code_mean_30days', 'parent_code_max_60days',
       'parent_code_min_60days', 'parent_code_median_60days', 'parent_code_mean_60days',
       'parent_code_max_180days', 'parent_code_min_180days', 'parent_code_median_180days',
       'parent_code_mean_180days']
parent_code.head(3)

parent_code  parent_code_total_sales_amount  parent_code_sales_7days  \
0          0.0                      18506893.0                5865675.0   
1          2.0                       1515784.0                 684725.0   
2          4.0                         44808.0                   1851.0   

   parent_code_ratio_7days  parent_code_sales_14days  \
0                    0.317                 8728702.0   
1                    0.452                  970749.0   
2                    0.041                    2631.0   

   parent_code_ratio_14days  parent_code_sales_30days  \
0                     0.472                12569867.0   
1                     0.640                 1187436.0   
2                     0.059                    4509.0   

   parent_code_ratio_30days  parent_code_sales_60days  \
0                     0.679                15237826.0   
1                     0.783                 1247623.0   
2                     0.101                   14144.0   

   parent_code_ratio_60days  parent_code_sales_180days  \
0                     0.823                 17362861.0   
1                     0.823                  1382325.0   
2                     0.316                    32940.0   

   parent_code_ratio_180days  parent_code_max_7days  parent_code_min_7days  \
0                      0.938                 9999.0                  100.0   
1                      0.912                 4038.0                  149.0   
2                      0.735                  858.0                  490.0   

   parent_code_median_7days  parent_code_mean_7days  parent_code_max_14days  \
0               1028.524461                   899.0                  9999.0   
1                541.712816                   449.0                  8076.0   
2                617.000000                   503.0                   858.0   

   parent_code_min_14days  parent_code_median_14days  parent_code_mean_14days  \
0                   100.0                1043.728566                    899.0   
1                   149.0                 555.030875                    458.0   
2                   390.0                 526.200000                    490.0   

   parent_code_max_30days  parent_code_min_30days  parent_code_median_30days  \
0                 16990.0                   100.0                1103.200544   
1                  8076.0                    99.0                 560.375649   
2                   939.0                   390.0                 644.142857   

   parent_code_mean_30days  parent_code_max_60days  parent_code_min_60days  \
0                    950.0                 16990.0                   100.0   
1                    458.0                  8076.0                    99.0   
2                    503.0                  1878.0                   390.0   

   parent_code_median_60days  parent_code_mean_60days  \
0                1132.418698                    989.0   
1                 561.739307                    458.0   
2                 884.000000                    898.5   

   parent_code_max_180days  parent_code_min_180days  \
0                  16990.0                    100.0   
1                   8076.0                     99.0   
2                   1878.0                    230.0   

   parent_code_median_180days  parent_code_mean_180days  
0                 1140.118261                     990.0  
1                  569.326606                     458.0  
2                  968.823529                     939.0

In [65]:
df=pd.merge(df,parent_code)
df.head(3)

business_hour_guid  product_launch_month  \
0  AB1BA373-E211-4206-A9F1-4A11F4DB5314                     7   
1  9584B6E4-264C-4FFC-8EF6-5223301E692A                     9   
2  124C1F67-488B-4B55-B449-C45650BDECC7                     4   

   product_launch_hour  product_launch_weekday   code  parent_code  \
0                    7                       3  536.0        536.0   
1                    7                       4  536.0        536.0   
2                    7                       1  536.0        536.0   

   is_Presales  is_Longtern  enable_isp  is_Discount  installment_3months  \
0            0            0           0            1                    1   
1            0            1           0            1                    1   
2            0            0           0            1                    1   

   installment_6months  installment_12months  deny_installment  \
0                    1                     1                 0   
1                    0                     0                 0   
2                    1                     1                 0   

   item_orig_price  item_price  item_default_daily_amount  \
0           3000.0      1088.0                         20   
1           1380.0       799.0                         20   
2           1718.0      1150.0                         20   

   quantity_multiplier  shippingdate  freight_amount  campagin_period  \
0                 10.0           0.0             0.0               21   
1                  1.0           0.0             0.0               81   
2                  1.0           0.0             0.0               21   

   order_deliver_days  unit_price  category_unit_price  \
0                   0       108.8                299.0   
1                   0       799.0                299.0   
2                   0      1150.0                299.0   

   compare_cate_unit_price  guid_subtotal  quantity_sold_guid  \
0                    0.364        16320.0               150.0   
1                    2.672         3995.0                 5.0   
2                    3.846         1150.0                 1.0   

   subtotal_allguid  quantity_sold_allguid  avg_unit_price  price_gap  \
0               0.0                    0.0           108.8      191.2   
1               0.0                    0.0           799.0      581.0   
2               0.0                    0.0          1150.0      568.0   

   hottimes  total_sales_amount  sales_7days  ratio_7days  sales_14days  \
0         0             16320.0       3264.0          0.2       11968.0   
1         0              3995.0        799.0          0.2         799.0   
2         0              1150.0       1150.0          1.0        1150.0   

   ratio_14days  sales_30days  ratio_30days  sales_60days  ratio_60days  \
0         0.733       16320.0           1.0       16320.0           1.0   
1         0.200         799.0           0.2        3196.0           0.8   
2         1.000        1150.0           1.0        1150.0           1.0   

   sales_180days  ratio_180days  code_total_sales_amount  code_sales_7days  \
0        16320.0            1.0              192491020.0        67838255.0   
1         3995.0            1.0              192491020.0        67838255.0   
2         1150.0            1.0              192491020.0        67838255.0   

   code_ratio_7days  code_sales_14days  code_ratio_14days  code_sales_30days  \
0             0.352        101498632.0              0.527        140193016.0   
1             0.352        101498632.0              0.527        140193016.0   
2             0.352        101498632.0              0.527        140193016.0   

   code_ratio_30days  code_sales_60days  code_ratio_60days  \
0              0.728        159588691.0              0.829   
1              0.728        159588691.0              0.829   
2              0.728        159588691.0              0.829   

   code_sales_180days  code_ratio_180days  code_max_7days  code_min_7days  \
0         18214011

## 觀察相關係數

In [66]:
corrtable=(df.corr()[df.corr().index=='sales_7days']).T
sales_7days=corrtable.sort_values(ascending=False,by='sales_7days')
sales_7days=sales_7days.drop(index=['quantity_sold_guid','ratio_7days','ratio_14days',
                                    'ratio_30days','ratio_60days','ratio_180days',
                                    'guid_subtotal','total_sales_amount','sales_7days',
                                    'sales_14days','sales_30days','sales_60days','sales_180days',
                                    'total_sales_amount'])
sales_7days

sales_7days
price_gap                     0.092855
unit_price                    0.087693
avg_unit_price                0.087487
item_price                    0.081142
item_orig_price               0.078186
compare_cate_unit_price       0.052588
code_max_60days               0.049080
code_max_30days               0.045493
deny_installment              0.039579
code_median_7days             0.039174
code_median_14days            0.039166
code_median_60days            0.038482
code_median_30days            0.038290
code_ratio_60days             0.038087
code_median_180days           0.038078
is_Discount                   0.037400
code_ratio_180days            0.037366
subtotal_allguid              0.037274
code_max_14days               0.034723
code_ratio_30days             0.032530
code_max_7days                0.031029
code_max_180days              0.029754
code_mean_14days              0.028963
code_mean_60days              0.028844
parent_code_max_14days        0.028777
code_mean_7days               0.028737
parent_code_max_30days        0.028594
parent_code_max_60days        0.028147
product_launch_hour           0.027990
freight_amount                0.027857
...                                ...
parent_code_median_7days     -0.001455
item_default_daily_amount    -0.001686
parent_code_median_14days    -0.001951
parent_code_mean_14days      -0.003501
installment_12months         -0.004238
parent_code_mean_7days       -0.004732
quantity_multiplier          -0.007283
code                         -0.007431
code_min_7days               -0.007519
code_min_180days             -0.008312
code_min_60days              -0.008835
code_min_30days              -0.009766
code_min_14days              -0.009883
hottimes                     -0.014831
parent_code_ratio_7days      -0.015863
parent_code_min_7days        -0.016447
parent_code_min_180days      -0.016790
parent_code_min_60days       -0.017069
parent_code_min_30days       -0.017208
parent_code_ratio_14days     -0.017875
parent_code_min_14days       -0.018491
parent_code_ratio_60days     -0.019347
parent_code_ratio_180days    -0.020417
parent_code                  -0.020447
parent_code_ratio_30days     -0.020465
product_launch_month         -0.030157
enable_isp                   -0.035241
installment_3months          -0.064589
campagin_period              -0.070708
is_Longtern                  -0.079576

[91 rows x 1 columns]

In [67]:
corrtable=(df.corr()[df.corr().index=='sales_14days']).T
sales_14days=corrtable.sort_values(ascending=False,by='sales_14days')
sales_14days=sales_14days.drop(index=['quantity_sold_guid','ratio_7days','ratio_14days',
                                    'ratio_30days','ratio_60days','ratio_180days',
                                    'guid_subtotal','total_sales_amount','sales_7days',
                                    'sales_14days','sales_30days','sales_60days','sales_180days',
                                    'total_sales_amount'])
sales_14days

sales_14days
price_gap                      0.100835
avg_unit_price                 0.095470
unit_price                     0.095354
item_price                     0.091266
item_orig_price                0.087832
compare_cate_unit_price        0.057756
code_max_60days                0.053803
code_median_14days             0.049861
code_max_30days                0.049501
code_median_7days              0.049326
code_median_60days             0.048874
code_median_30days             0.048768
code_median_180days            0.048456
subtotal_allguid               0.039111
code_mean_14days               0.038942
code_mean_60days               0.038690
is_Discount                    0.038683
code_mean_7days                0.038104
code_mean_180days              0.037691
code_mean_30days               0.037283
code_max_14days                0.036699
code_ratio_60days              0.036370
code_ratio_180days             0.036215
parent_code_max_30days         0.034549
parent_code_max_14days         0.034515
parent_code_max_60days         0.034315
parent_code_max_7days          0.032952
category_unit_price            0.032924
code_max_7days                 0.032684
freight_amount                 0.031586
...                                 ...
code_min_180days              -0.002327
code_min_60days               -0.002905
code_min_7days                -0.003555
code_min_30days               -0.003911
code_min_14days               -0.004009
installment_12months          -0.004314
parent_code_sales_180days     -0.004532
parent_code_sales_60days      -0.005529
quantity_multiplier           -0.005644
parent_code_sales_30days      -0.005933
parent_code_sales_14days      -0.006342
parent_code_sales_7days       -0.006542
code                          -0.006966
parent_code_min_7days         -0.012018
parent_code_min_60days        -0.012229
parent_code_min_30days        -0.012258
parent_code_min_180days       -0.012790
parent_code_min_14days        -0.014636
hottimes                      -0.016248
parent_code                   -0.018895
product_launch_month          -0.021423
parent_code_ratio_7days       -0.025444
parent_code_ratio_14days      -0.026898
parent_code_ratio_60days      -0.027387
parent_code_ratio_180days     -0.027678
parent_code_ratio_30days      -0.029358
enable_isp                    -0.034718
installment_3months           -0.060495
campagin_period               -0.068389
is_Longtern                   -0.069423

[91 rows x 1 columns]

In [68]:
corrtable=(df.corr()[df.corr().index=='sales_30days']).T
sales_30days=corrtable.sort_values(ascending=False,by='sales_30days')
sales_30days=sales_30days.drop(index=['quantity_sold_guid','ratio_7days','ratio_14days',
                                    'ratio_30days','ratio_60days','ratio_180days',
                                    'guid_subtotal','total_sales_amount','sales_7days',
                                    'sales_14days','sales_30days','sales_60days','sales_180days',
                                    'total_sales_amount'])
sales_30days

sales_30days
price_gap                           0.106340
avg_unit_price                      0.105410
unit_price                          0.105327
item_price                          0.103596
item_orig_price                     0.096455
compare_cate_unit_price             0.065042
code_median_14days                  0.060582
code_median_30days                  0.059866
code_median_7days                   0.059829
code_median_60days                  0.059735
code_median_180days                 0.059145
code_max_60days                     0.056398
code_max_30days                     0.051415
code_mean_60days                    0.050457
code_mean_14days                    0.050218
code_mean_180days                   0.049415
code_mean_7days                     0.049273
code_mean_30days                    0.049214
subtotal_allguid                    0.045179
parent_code_max_60days              0.041995
parent_code_max_30days              0.041769
parent_code_max_14days              0.041491
parent_code_max_7days               0.039835
category_unit_price                 0.039376
is_Discount                         0.039237
code_max_14days                     0.035891
freight_amount                      0.033358
code_max_180days                    0.032651
code_max_7days                      0.031857
code_ratio_180days                  0.031423
...                                      ...
product_launch_weekday              0.001683
code_min_7days                      0.001641
item_default_daily_amount          -0.001502
quantity_multiplier                -0.002480
code_ratio_7days                   -0.003841
installment_12months               -0.004727
code                               -0.006384
parent_code_min_60days             -0.007367
parent_code_min_30days             -0.007411
parent_code_min_7days              -0.008034
parent_code_min_180days            -0.008189
parent_code_total_sales_amount     -0.009636
product_launch_month               -0.010248
parent_code_min_14days             -0.010983
parent_code_sales_180days          -0.012082
parent_code_sales_60days           -0.013238
parent_code_sales_30days           -0.013743
parent_code_sales_14days           -0.014497
parent_code_sales_7days            -0.014970
parent_code                        -0.015728
hottimes                           -0.020044
enable_isp                         -0.032437
parent_code_ratio_180days          -0.035402
is_Longtern                        -0.035597
parent_code_ratio_60days           -0.035875
parent_code_ratio_7days            -0.036148
parent_code_ratio_14days           -0.036793
parent_code_ratio_30days           -0.038623
installment_3months                -0.048713
campagin_period                    -0.058042

[91 rows x 1 columns]

In [69]:
corrtable=(df.corr()[df.corr().index=='sales_60days']).T
sales_60days=corrtable.sort_values(ascending=False,by='sales_60days')
sales_60days=sales_60days.drop(index=['quantity_sold_guid','ratio_7days','ratio_14days',
                                    'ratio_30days','ratio_60days','ratio_180days',
                                    'guid_subtotal','total_sales_amount','sales_7days',
                                    'sales_14days','sales_30days','sales_60days','sales_180days',
                                    'total_sales_amount'])
sales_60days

sales_60days
unit_price                          0.114531
avg_unit_price                      0.114444
item_price                          0.113826
price_gap                           0.108478
item_orig_price                     0.103314
code_median_14days                  0.072308
code_median_30days                  0.071681
code_median_60days                  0.071558
code_median_7days                   0.071362
code_median_180days                 0.070914
compare_cate_unit_price             0.066843
code_mean_60days                    0.063476
code_mean_14days                    0.062712
code_mean_180days                   0.062535
code_mean_30days                    0.062151
code_mean_7days                     0.061398
code_max_60days                     0.054024
category_unit_price                 0.052442
code_max_30days                     0.047704
subtotal_allguid                    0.044677
deny_installment                    0.038468
parent_code_max_60days              0.037962
is_Discount                         0.037921
parent_code_max_30days              0.037022
parent_code_max_14days              0.036353
shippingdate                        0.035027
parent_code_max_7days               0.034875
code_max_14days                     0.033169
code_max_180days                    0.031326
code_max_7days                      0.030078
...                                      ...
is_Longtern                         0.007248
product_launch_weekday              0.002077
parent_code_min_60days              0.000598
parent_code_min_30days              0.000483
parent_code_min_180days            -0.000157
parent_code_min_7days              -0.000434
quantity_multiplier                -0.000507
item_default_daily_amount          -0.001147
product_launch_month               -0.002215
parent_code_min_14days             -0.003711
code                               -0.005500
code_ratio_14days                  -0.006760
installment_12months               -0.007508
parent_code_total_sales_amount     -0.011479
parent_code                        -0.011685
parent_code_sales_180days          -0.013575
parent_code_sales_60days           -0.014661
parent_code_sales_30days           -0.015211
parent_code_sales_14days           -0.016065
code_ratio_7days                   -0.016346
parent_code_sales_7days            -0.016690
hottimes                           -0.021456
enable_isp                         -0.026781
parent_code_ratio_180days          -0.031971
parent_code_ratio_60days           -0.033509
parent_code_ratio_14days           -0.035775
parent_code_ratio_7days            -0.035829
campagin_period                    -0.036757
parent_code_ratio_30days           -0.037260
installment_3months                -0.037571

[91 rows x 1 columns]

In [70]:
corrtable=(df.corr()[df.corr().index=='sales_180days']).T
sales_180days=corrtable.sort_values(ascending=False,by='sales_180days')
sales_180days=sales_180days.drop(index=['quantity_sold_guid','ratio_7days','ratio_14days',
                                    'ratio_30days','ratio_60days','ratio_180days',
                                    'guid_subtotal','total_sales_amount','sales_7days',
                                    'sales_14days','sales_30days','sales_60days','sales_180days',
                                    'total_sales_amount'])
sales_180days

sales_180days
item_price                           0.117506
avg_unit_price                       0.117474
unit_price                           0.117419
price_gap                            0.107218
item_orig_price                      0.106314
code_median_14days                   0.084806
code_median_30days                   0.083953
code_median_60days                   0.083722
code_median_7days                    0.083392
code_median_180days                  0.083207
code_mean_60days                     0.077912
code_mean_180days                    0.077281
code_mean_14days                     0.077006
code_mean_30days                     0.076459
code_mean_7days                      0.074894
category_unit_price                  0.068878
compare_cate_unit_price              0.062214
code_max_60days                      0.046533
subtotal_allguid                     0.046281
deny_installment                     0.041411
is_Longtern                          0.039858
code_max_30days                      0.038846
is_Discount                          0.036305
code_min_60days                      0.035620
code_min_180days                     0.035219
code_min_30days                      0.034486
shippingdate                         0.034478
code_min_14days                      0.033626
parent_code_mean_30days              0.030758
parent_code_mean_180days             0.030277
...                                       ...
code_sales_30days                    0.006846
product_launch_weekday               0.004533
quantity_multiplier                  0.003133
code_sales_14days                    0.002713
product_launch_month                 0.002185
code_ratio_180days                   0.001639
campagin_period                      0.001429
item_default_daily_amount           -0.000867
code_sales_7days                    -0.000940
code_ratio_60days                   -0.003929
code                                -0.004218
parent_code                         -0.004620
installment_12months                -0.006517
parent_code_total_sales_amount      -0.011591
parent_code_sales_180days           -0.013023
code_ratio_30days                   -0.013481
parent_code_sales_60days            -0.013979
parent_code_sales_30days            -0.014576
parent_code_sales_14days            -0.015516
parent_code_sales_7days             -0.016341
enable_isp                          -0.017572
hottimes                            -0.021827
parent_code_ratio_180days           -0.023600
installment_3months                 -0.025314
code_ratio_14days                   -0.026009
parent_code_ratio_60days            -0.026890
parent_code_ratio_14days            -0.030706
parent_code_ratio_30days            -0.031742
parent_code_ratio_7days             -0.031920
code_ratio_7days                    -0.033699

[91 rows x 1 columns]

## 因為code的類別數較大,故使用計數編碼

In [71]:
parent_code=df.parent_code.value_counts().to_frame()
parent_code=parent_code.rename(columns={'parent_code':'parent_code_counts'})
parent_code.index.name='parent_code'
#code.columns=[['code','code_counts']]
parent_code=parent_code.reset_index(drop=False)
df=pd.merge(df,parent_code,how='left')
df=df.drop(columns=['parent_code'])

df.head(3)

business_hour_guid  product_launch_month  \
0  AB1BA373-E211-4206-A9F1-4A11F4DB5314                     7   
1  9584B6E4-264C-4FFC-8EF6-5223301E692A                     9   
2  124C1F67-488B-4B55-B449-C45650BDECC7                     4   

   product_launch_hour  product_launch_weekday   code  is_Presales  \
0                    7                       3  536.0            0   
1                    7                       4  536.0            0   
2                    7                       1  536.0            0   

   is_Longtern  enable_isp  is_Discount  installment_3months  \
0            0           0            1                    1   
1            1           0            1                    1   
2            0           0            1                    1   

   installment_6months  installment_12months  deny_installment  \
0                    1                     1                 0   
1                    0                     0                 0   
2                    1                     1                 0   

   item_orig_price  item_price  item_default_daily_amount  \
0           3000.0      1088.0                         20   
1           1380.0       799.0                         20   
2           1718.0      1150.0                         20   

   quantity_multiplier  shippingdate  freight_amount  campagin_period  \
0                 10.0           0.0             0.0               21   
1                  1.0           0.0             0.0               81   
2                  1.0           0.0             0.0               21   

   order_deliver_days  unit_price  category_unit_price  \
0                   0       108.8                299.0   
1                   0       799.0                299.0   
2                   0      1150.0                299.0   

   compare_cate_unit_price  guid_subtotal  quantity_sold_guid  \
0                    0.364        16320.0               150.0   
1                    2.672         3995.0                 5.0   
2                    3.846         1150.0                 1.0   

   subtotal_allguid  quantity_sold_allguid  avg_unit_price  price_gap  \
0               0.0                    0.0           108.8      191.2   
1               0.0                    0.0           799.0      581.0   
2               0.0                    0.0          1150.0      568.0   

   hottimes  total_sales_amount  sales_7days  ratio_7days  sales_14days  \
0         0             16320.0       3264.0          0.2       11968.0   
1         0              3995.0        799.0          0.2         799.0   
2         0              1150.0       1150.0          1.0        1150.0   

   ratio_14days  sales_30days  ratio_30days  sales_60days  ratio_60days  \
0         0.733       16320.0           1.0       16320.0           1.0   
1         0.200         799.0           0.2        3196.0           0.8   
2         1.000        1150.0           1.0        1150.0           1.0   

   sales_180days  ratio_180days  code_total_sales_amount  code_sales_7days  \
0        16320.0            1.0              192491020.0        67838255.0   
1         3995.0            1.0              192491020.0        67838255.0   
2         1150.0            1.0              192491020.0        67838255.0   

   code_ratio_7days  code_sales_14days  code_ratio_14days  code_sales_30days  \
0             0.352        101498632.0              0.527        140193016.0   
1             0.352        101498632.0              0.527        140193016.0   
2             0.352        101498632.0              0.527        140193016.0   

   code_ratio_30days  code_sales_60days  code_ratio_60days  \
0              0.728        159588691.0              0.829   
1              0.728        159588691.0              0.829   
2              0.728        159588691.0              0.829   

   code_sales_180days  code_ratio_180days  code_max_7days  code_min_7days  \
0         182140117.0               0.946         31194.0            1

In [72]:
code=df.code.value_counts().to_frame()
code=code.rename(columns={'code':'code_counts'})
code.index.name='code'
#code.columns=[['code','code_counts']]
code=code.reset_index(drop=False)
df=pd.merge(df,code,how='left')

df=df.drop(columns=['code'])
df.head()

business_hour_guid  product_launch_month  \
0  AB1BA373-E211-4206-A9F1-4A11F4DB5314                     7   
1  9584B6E4-264C-4FFC-8EF6-5223301E692A                     9   
2  124C1F67-488B-4B55-B449-C45650BDECC7                     4   
3  CB632369-0836-46C3-9DBE-2B6695301885                    11   
4  9D944ADD-70C5-4A16-8318-F03C80B72A1B                     4   

   product_launch_hour  product_launch_weekday  is_Presales  is_Longtern  \
0                    7                       3            0            0   
1                    7                       4            0            1   
2                    7                       1            0            0   
3                    7                       0            0            0   
4                    7                       3            0            0   

   enable_isp  is_Discount  installment_3months  installment_6months  \
0           0            1                    1                    1   
1           0            1                    1                    0   
2           0            1                    1                    1   
3           0            1                    1                    1   
4           0            1                    0                    0   

   installment_12months  deny_installment  item_orig_price  item_price  \
0                     1                 0           3000.0      1088.0   
1                     0                 0           1380.0       799.0   
2                     1                 0           1718.0      1150.0   
3                     1                 0            498.0       249.0   
4                     0                 0            998.0       370.0   

   item_default_daily_amount  quantity_multiplier  shippingdate  \
0                         20                 10.0           0.0   
1                         20                  1.0           0.0   
2                         20                  1.0           0.0   
3                         20                  2.0           0.0   
4                         20                  2.0           0.0   

   freight_amount  campagin_period  order_deliver_days  unit_price  \
0             0.0               21                   0       108.8   
1             0.0               81                   0       799.0   
2             0.0               21                   0      1150.0   
3             0.0               21                   0       124.5   
4             0.0               21                   0       185.0   

   category_unit_price  compare_cate_unit_price  guid_subtotal  \
0                299.0                    0.364        16320.0   
1                299.0                    2.672         3995.0   
2                299.0                    3.846         1150.0   
3                299.0                    0.416          498.0   
4                299.0                    0.619         1110.0   

   quantity_sold_guid  subtotal_allguid  quantity_sold_allguid  \
0               150.0               0.0                    0.0   
1                 5.0               0.0                    0.0   
2                 1.0               0.0                    0.0   
3                 4.0               0.0                    0.0   
4                 6.0               0.0                    0.0   

   avg_unit_price  price_gap  hottimes  total_sales_amount  sales_7days  \
0           108.8      191.2         0             16320.0       3264.0   
1           799.0      581.0         0              3995.0        799.0   
2          1150.0      568.0         0              1150.0       1150.0   
3           124.5      124.5         0               498.0        498.0   
4           185.0      314.0         0              1110.0          0.0   

   ratio_7days  sales_14days  ratio_14days  sales_30days  ratio_30days  \
0          0.2       11968.0         0.733       16320.0           1.0   
1          0.2         799.0         0.200         799.0           0.2   
2   

In [73]:
df['orig_unit_price']=df['item_orig_price']/df['quantity_multiplier']

## 預測銷量

In [96]:
continousIndex=['item_orig_price','item_price','item_default_daily_amount',
                'subtotal_allguid','quantity_sold_allguid',
                #'ratio_7days','ratio_14days','ratio_30days','ratio_60days','ratio_180days'
                #'code_ratio_7days','code_ratio_14days','code_ratio_30days','code_ratio_60days','code_ratio_180days',
                'code_total_sales_amount',
                'code_sales_7days','code_sales_14days','code_sales_30days','code_sales_60days','code_sales_180days',
                'code_max_7days','code_min_7days','code_median_7days','code_mean_7days','code_max_14days','code_min_14days',
                'code_median_14days','code_mean_14days','code_max_30days','code_min_30days','code_median_30days',
                'code_mean_30days','code_max_60days','code_min_60days','code_median_60days','code_mean_60days',
                'code_max_180days','code_min_180days','code_median_180days','code_mean_180days','parent_code_total_sales_amount',
                #'parent_code_ratio_7days','parent_code_ratio_14days','parent_code_ratio_30days','parent_code_ratio_60days','parent_code_ratio_180days',
                'parent_code_sales_7days','parent_code_sales_14days','parent_code_sales_30days','parent_code_sales_60days',
                'parent_code_sales_180days','parent_code_max_7days','parent_code_min_7days',
                'parent_code_median_7days','parent_code_mean_7days','parent_code_max_14days','parent_code_min_14days',
                'parent_code_median_14days','parent_code_mean_14days','parent_code_max_30days','parent_code_min_30days',
                'parent_code_median_30days','parent_code_mean_30days','parent_code_max_60days','parent_code_min_60days',
                'parent_code_median_60days','parent_code_mean_60days','parent_code_max_180days','parent_code_min_180days',
                'parent_code_median_180days','parent_code_mean_180days','parent_code_counts','code_counts','orig_unit_price']

targetIndex=['guid_subtotal','sales_7days','sales_14days','sales_30days', 'sales_60days','sales_180days']
categoryIndex=['product_launch_month','product_launch_hour','product_launch_weekday','is_Presales','is_Longtern',
'enable_isp','is_Discount','installment_3months','installment_6months','installment_12months','deny_installment']

In [178]:
##將銷量相關取log
df_new=df.drop(columns='business_hour_guid')
ohe_df=pd.get_dummies(df_new,columns=categoryIndex)
ohe_df[continousIndex] = np.log1p(ohe_df[continousIndex])
ohe_df[targetIndex] = np.log1p(ohe_df[targetIndex])

In [248]:
from sklearn.metrics import mean_squared_error, r2_score,mean_absolute_error
def trainTestValSplit(df):
    # 將資料切分為訓練(80%),測試(10%)和驗證(10%)三份。
    train=df.sample(frac=0.8,random_state=0)
    test=df.drop(train.index )
    val=test.sample(frac=0.5,random_state=0)
    test=test.drop(val.index)
    return train,test,val

def dfXYSplit(df,targetIndex,targetName):
    # 將特徵和目標變數切成兩份資料。
    
    dfX=df.drop(columns=targetIndex)
    dfY=df[targetName]
    
    return dfX,dfY
def trainmodel(df,targetIndex,targetname,model):
    train,test,val=trainTestValSplit(df)
    trainX,trainY=dfXYSplit(train,targetIndex,targetname)
    testX,testY=dfXYSplit(test,targetIndex,targetname)
    valX,valY=dfXYSplit(val,targetIndex,targetname)
    model.fit(trainX,trainY)
    testPredY=model.predict(testX)
    valPredY=model.predict(valX)
    print(model)
    print("val_mean squared error: %.2f"
      % mean_squared_error(valY, valPredY))

    print("val_mean_absolute_error: %.2f"
      % mean_absolute_error(valY, valPredY))
    print("val_r2_score: %.2f"
      % r2_score(valY,valPredY))
    subtotal=pd.DataFrame()
    subtotal['predict']=np.expm1(np.array(testPredY.tolist()).reshape(-1))
    subtotal['real']=np.expm1(testY[testY.columns[0]].reset_index(drop=True))
    def check(x):
        if x <1 :
            return 0
        else:
            return x
    subtotal.predict=subtotal.predict.apply(check)

    print("test_mean squared error: %.2f"
      % mean_squared_error(subtotal['real'],subtotal['predict']))

    print("test_mean_absolute_error: %.2f"
      % mean_absolute_error(subtotal['real'],subtotal['predict']))
    print("test_r2_score: %.2f"
      % r2_score(subtotal['real'],subtotal['predict']))
    return subtotal

## DecisionTree

In [249]:
from sklearn.tree import DecisionTreeRegressor
model = DecisionTreeRegressor(random_state=0)
targetname=['sales_7days']
dt=trainmodel(ohe_df,targetIndex,targetname,model)
dt.head(6)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=1,
           min_samples_split=2, min_weight_fraction_leaf=0.0,
           presort=False, random_state=0, splitter='best')
val_mean squared error: 0.00
val_mean_absolute_error: 0.01
val_r2_score: 1.00
test_mean squared error: 1941433.53
test_mean_absolute_error: 80.04
test_r2_score: 0.97


predict     real
0   1199.0   1199.0
1   1298.0   1302.0
2      0.0      0.0
3      0.0      0.0
4   1797.0   1797.0
5  10850.0  10746.0

In [251]:
model = DecisionTreeRegressor(random_state=0)
targetname=['sales_14days']
dt=trainmodel(ohe_df,targetIndex,targetname,model)
dt.head(6)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=1,
           min_samples_split=2, min_weight_fraction_leaf=0.0,
           presort=False, random_state=0, splitter='best')
val_mean squared error: 0.00
val_mean_absolute_error: 0.01
val_r2_score: 1.00
test_mean squared error: 19436091.10
test_mean_absolute_error: 134.19
test_r2_score: 0.90


predict     real
0   1199.0   1199.0
1   3897.0   3906.0
2    420.0    420.0
3      0.0      0.0
4   2396.0   2396.0
5  18522.0  17910.0

In [252]:
model = DecisionTreeRegressor(random_state=0)
targetname=['sales_30days']
dt=trainmodel(ohe_df,targetIndex,targetname,model)
dt.head(6)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=1,
           min_samples_split=2, min_weight_fraction_leaf=0.0,
           presort=False, random_state=0, splitter='best')
val_mean squared error: 0.00
val_mean_absolute_error: 0.00
val_r2_score: 1.00
test_mean squared error: 24940602.11
test_mean_absolute_error: 169.24
test_r2_score: 0.95


predict     real
0   2396.655736   2398.0
1   6510.000000   6510.0
2   1260.009705   1260.0
3      0.000000      0.0
4   2994.243553   2995.0
5  27453.570170  27462.0

In [253]:
model = DecisionTreeRegressor(random_state=0)
targetname=['sales_60days']
dt=trainmodel(ohe_df,targetIndex,targetname,model)
dt.head(6)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=1,
           min_samples_split=2, min_weight_fraction_leaf=0.0,
           presort=False, random_state=0, splitter='best')
val_mean squared error: 0.00
val_mean_absolute_error: 0.00
val_r2_score: 1.00
test_mean squared error: 17610517.29
test_mean_absolute_error: 157.08
test_r2_score: 0.98


predict     real
0   2398.329222   2398.0
1   6510.000000   6510.0
2   1259.921225   1260.0
3    899.000000    878.0
4   2994.245181   2995.0
5  27459.999781  27462.0

In [254]:
model = DecisionTreeRegressor(random_state=0)
targetname=['sales_180days']
dt=trainmodel(ohe_df,targetIndex,targetname,model)
dt.head(6)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=1,
           min_samples_split=2, min_weight_fraction_leaf=0.0,
           presort=False, random_state=0, splitter='best')
val_mean squared error: 0.00
val_mean_absolute_error: 0.00
val_r2_score: 1.00
test_mean squared error: 9644652.55
test_mean_absolute_error: 109.43
test_r2_score: 0.99


predict     real
0   2398.321794   2398.0
1   6510.555468   6510.0
2   1259.909053   1260.0
3   4288.999883   4390.0
4   2994.778841   2995.0
5  27471.570160  27462.0

### XGBoost

In [211]:
import os
os.environ['KMP_DUPLICATE_LIB_OK']='True'
import xgboost as xgb
#model = MultiOutputRegressor(xgb.XGBRegressor(n_estimators=200)).fit(x_train, y_train)
model = xgb.XGBRegressor(n_estimators=100)
targetname=['sales_7days']
xgb=trainmodel(ohe_df,targetIndex,targetname,model)
xgb.head(6)

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
       n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=True, subsample=1)
val_mean squared error: 0.01
val_mean_absolute_error: 0.04
val_r2_score: 1.00
test_mean squared error: 8868449.63
test_mean_absolute_error: 194.04
test_r2_score: 0.93


predict    real
0   494.992167   449.0
1     0.000000     0.0
2     0.000000     0.0
3  4615.099902  5196.0
4  1646.930730  1500.0
5     0.000000     0.0

In [212]:
import os
os.environ['KMP_DUPLICATE_LIB_OK']='True'
import xgboost as xgb
#model = MultiOutputRegressor(xgb.XGBRegressor(n_estimators=200)).fit(x_train, y_train)
model = xgb.XGBRegressor(n_estimators=100)
targetname=['sales_14days']
xgb=trainmodel(ohe_df,targetIndex,targetname,model)
xgb.head(6)

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
       n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=True, subsample=1)
val_mean squared error: 0.00
val_mean_absolute_error: 0.04
val_r2_score: 1.00
test_mean squared error: 18059198.55
test_mean_absolute_error: 310.52
test_r2_score: 0.95


predict    real
0     0.000000     0.0
1     0.000000     0.0
2     0.000000     0.0
3  3425.555598  3408.0
4   249.900000   250.0
5   553.851126   537.0

In [213]:
import os
os.environ['KMP_DUPLICATE_LIB_OK']='True'
import xgboost as xgb
#model = MultiOutputRegressor(xgb.XGBRegressor(n_estimators=200)).fit(x_train, y_train)
model = xgb.XGBRegressor(n_estimators=100)
targetname=['sales_30days']
xgb=trainmodel(ohe_df,targetIndex,targetname,model)
xgb.head(6)

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
       n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=True, subsample=1)
val_mean squared error: 0.00
val_mean_absolute_error: 0.03
val_r2_score: 1.00
test_mean squared error: 33522000.34
test_mean_absolute_error: 350.15
test_r2_score: 0.97


predict     real
0   1009.376129   1043.0
1   3054.851793   3152.0
2    501.427778    490.0
3   3031.293628   2994.0
4   8637.398390   8670.0
5  14196.681150  15181.0

In [214]:
import os
os.environ['KMP_DUPLICATE_LIB_OK']='True'
import xgboost as xgb
#model = MultiOutputRegressor(xgb.XGBRegressor(n_estimators=200)).fit(x_train, y_train)
model = xgb.XGBRegressor(n_estimators=100)
targetname=['sales_60days']
xgb=trainmodel(ohe_df,targetIndex,targetname,model)
xgb.head(6)

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
       n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=True, subsample=1)
val_mean squared error: 0.00
val_mean_absolute_error: 0.02
val_r2_score: 1.00
test_mean squared error: 101946159.41
test_mean_absolute_error: 384.31
test_r2_score: 0.94


predict    real
0  1167.800281  1150.0
1   595.165426   596.0
2  1893.517658  1880.0
3  2219.887679  2394.0
4     0.000000     0.0
5  2652.704361  2748.0

In [215]:
import os
os.environ['KMP_DUPLICATE_LIB_OK']='True'
import xgboost as xgb
#model = MultiOutputRegressor(xgb.XGBRegressor(n_estimators=200)).fit(x_train, y_train)
model = xgb.XGBRegressor(n_estimators=100)
targetname=['sales_180days']
xgb=trainmodel(ohe_df,targetIndex,targetname,model)
xgb.head(6)

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
       n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=True, subsample=1)
val_mean squared error: 0.00
val_mean_absolute_error: 0.02
val_r2_score: 1.00
test_mean squared error: 33614690.40
test_mean_absolute_error: 318.67
test_r2_score: 0.99


predict    real
0   495.539037   498.0
1     0.000000     0.0
2   200.419037   199.0
3   239.278266   245.0
4  3236.496973  3240.0
5  1175.714251  1192.0

## RandomForest

In [216]:
from sklearn.ensemble import RandomForestRegressor
model=RandomForestRegressor(max_depth=2, random_state=0,n_estimators=100)
targetname=['sales_7days']
rf=trainmodel(ohe_df,targetIndex,targetname,model)
rf.head(6)

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:22: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().


RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=2,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=100, n_jobs=None,
           oob_score=False, random_state=0, verbose=0, warm_start=False)
val_mean squared error: 0.48
val_mean_absolute_error: 0.44
val_r2_score: 0.96
test_mean squared error: 146949792.97
test_mean_absolute_error: 1954.72
test_r2_score: 0.05


predict      real
0   916.144735    1298.0
1  3065.220022    1098.0
2     0.000000       0.0
3  1035.412975    3195.0
4     0.000000       0.0
5  4317.177573  592752.0

## LinearRegression

In [217]:
from sklearn.linear_model import LinearRegression
model=LinearRegression()
targetname=['sales_7days']
lr=trainmodel(ohe_df,targetIndex,targetname,model)
lr.head(6)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)
val_mean squared error: 5.11
val_mean_absolute_error: 1.94
val_r2_score: 0.62
test_mean squared error: 151504274943033536.00
test_mean_absolute_error: 3034532.50
test_r2_score: -1271478634.72


predict     real
0    24.802337   1098.0
1     4.369510      0.0
2     0.000000      0.0
3  1067.895627  10080.0
4     4.179088      0.0
5     1.271104      0.0

## Ridge 

In [218]:
from sklearn.linear_model import Lasso, Ridge
model=Ridge()
targetname=['sales_7days']
rl=trainmodel(ohe_df,targetIndex,targetname,model)
rl.head(6)

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)
val_mean squared error: 5.03
val_mean_absolute_error: 1.93
val_r2_score: 0.62
test_mean squared error: 22048781929090.02
test_mean_absolute_error: 41907.71
test_r2_score: -187456.65


predict    real
0   162.723505  2457.0
1  2467.404631  3195.0
2     1.301394     0.0
3   372.523946  1112.0
4     0.000000     0.0
5  1132.480334  1797.0

## 將資料切分訓練(80%),測試(10%)和驗證(10%)三份


* 未作細調參數前的MAE R2整理：

Model                    | Val_MAE                  |Val_R2                  | 
----------------         |:---------:               |---------:              | 
DecisionTreeRegressor    | 0.01/0.01/0.00/0.00/0.00 |1.0/1.0/1.0/1.0/1.0     |    0 | 
XGBRegressor             | 0.04/0.04/0.03/0.02/0.02 |1.0/1.0/1.0/1.0/1.0     |    0 |  
RandomForestRegressor    | 0.44                     |0.96                    |    0 |  
LinearRegression         | 1.94                     |0.62                    |    0 |  
Ridge                    | 1.93                     |0.62                    |    0 |  


* 因為所有數值型變數都有經過對數轉換，故在檢查測試集還原log後的表現，選擇以DecisionTree 與XGBregressor 進行參數調教

## 使用網格搜尋法
* 將資料重新分為訓練集與測試集

In [223]:
from sklearn import datasets, metrics
from sklearn.model_selection import train_test_split, KFold, GridSearchCV
targetname=['sales_7days']
trainX,trainY=dfXYSplit(ohe_df,targetIndex,targetname)
x_train, x_test, y_train, y_test = train_test_split(trainX,trainY, test_size=0.2, random_state=4)

In [230]:
import os
os.environ['KMP_DUPLICATE_LIB_OK']='True'
import xgboost as xgb
model =  xgb.XGBRegressor( )


# 設定要訓練的超參數組合
n_estimators = [10,25,50, 100, 150]
max_depth = [1, 3, 5,7,10]
param_grid = dict(n_estimators=n_estimators, max_depth=max_depth)

## 建立搜尋物件，放入模型及參數組合字典 (n_jobs=-1 會使用全部 cpu 平行運算)
grid_search = GridSearchCV(model, param_grid,n_jobs=-1, verbose=1)

# 開始搜尋最佳參數
grid_result = grid_search.fit(x_train,y_train)
print("Best Accuracy: %f using %s" % (grid_result.best_score_, grid_result.best_params_))


Fitting 3 folds for each of 25 candidates, totalling 75 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
/anaconda3/lib/python3.6/site-packages/sklearn/externals/joblib/externals/loky/process_executor.py:706: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  "timeout or by a memory leak.", UserWarning
[Parallel(n_jobs=-1)]: Done  42 tasks      | elapsed: 11.1min
/anaconda3/lib/python3.6/site-packages/sklearn/externals/joblib/externals/loky/process_executor.py:706: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  "timeout or by a memory leak.", UserWarning
/anaconda3/lib/python3.6/site-packages/sklearn/externals/joblib/externals/loky/process_executor.py:706: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  "timeout or by a memory 

KeyboardInterrupt: 

In [234]:
import os
os.environ['KMP_DUPLICATE_LIB_OK']='True'
model =  DecisionTreeRegressor()

# 設定要訓練的超參數組合
min_samples_split=[2,4,6,8,10]
min_samples_leaf=[1, 3,5,7,9]
max_depth = [1,3, 5,10,15,20]
param_grid = dict(max_depth=max_depth,min_samples_split=min_samples_split,min_samples_leaf=min_samples_leaf)

## 建立搜尋物件，放入模型及參數組合字典 (n_jobs=-1 會使用全部 cpu 平行運算)
grid_search = GridSearchCV(model, param_grid,n_jobs=-1, verbose=1)

# 開始搜尋最佳參數
grid_result = grid_search.fit(x_train,y_train)
print("Best combination: %f using %s" % (grid_result.best_score_, grid_result.best_params_))


/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.


Fitting 3 folds for each of 150 candidates, totalling 450 fits


[Parallel(n_jobs=-1)]: Done  42 tasks      | elapsed:   14.4s
[Parallel(n_jobs=-1)]: Done 192 tasks      | elapsed:  1.5min
[Parallel(n_jobs=-1)]: Done 442 tasks      | elapsed:  7.3min
[Parallel(n_jobs=-1)]: Done 450 out of 450 | elapsed:  7.5min finished


Best combination: 0.999931 using {'max_depth': 20, 'min_samples_leaf': 3, 'min_samples_split': 8}


In [255]:
model = DecisionTreeRegressor(random_state=0,max_depth=20,min_samples_leaf=3,min_samples_split=8)
targetname=['sales_7days']
dt=trainmodel(ohe_df,targetIndex,targetname,model)
dt.head(6)

DecisionTreeRegressor(criterion='mse', max_depth=20, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=3,
           min_samples_split=8, min_weight_fraction_leaf=0.0,
           presort=False, random_state=0, splitter='best')
val_mean squared error: 0.00
val_mean_absolute_error: 0.01
val_r2_score: 1.00
test_mean squared error: 3785882.32
test_mean_absolute_error: 85.19
test_r2_score: 0.95


predict     real
0   1199.000000   1199.0
1   1308.464593   1302.0
2      0.000000      0.0
3      0.000000      0.0
4   1797.000000   1797.0
5  10788.848679  10746.0

In [260]:
model = DecisionTreeRegressor(random_state=0,max_depth=20,min_samples_leaf=3,min_samples_split=8)
targetname=['sales_14days']
dt=trainmodel(ohe_df,targetIndex,targetname,model)
dt.head(6)

DecisionTreeRegressor(criterion='mse', max_depth=20, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=3,
           min_samples_split=8, min_weight_fraction_leaf=0.0,
           presort=False, random_state=0, splitter='best')
val_mean squared error: 0.00
val_mean_absolute_error: 0.01
val_r2_score: 1.00
test_mean squared error: 9608770.75
test_mean_absolute_error: 114.42
test_r2_score: 0.95


predict     real
0   1199.000000   1199.0
1   3914.980848   3906.0
2    420.831591    420.0
3      0.000000      0.0
4   2396.000000   2396.0
5  18318.422564  17910.0

In [261]:
model = DecisionTreeRegressor(random_state=0,max_depth=20,min_samples_leaf=3,min_samples_split=8)
targetname=['sales_30days']
dt=trainmodel(ohe_df,targetIndex,targetname,model)
dt.head(6)

DecisionTreeRegressor(criterion='mse', max_depth=20, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=3,
           min_samples_split=8, min_weight_fraction_leaf=0.0,
           presort=False, random_state=0, splitter='best')
val_mean squared error: 0.00
val_mean_absolute_error: 0.00
val_r2_score: 1.00
test_mean squared error: 38738481.51
test_mean_absolute_error: 185.14
test_r2_score: 0.93


predict     real
0   2396.655736   2398.0
1   6508.142189   6510.0
2   1260.009705   1260.0
3      0.000000      0.0
4   2994.243553   2995.0
5  27457.999534  27462.0

In [258]:
model = DecisionTreeRegressor(random_state=0,max_depth=20,min_samples_leaf=3,min_samples_split=8)
targetname=['sales_60days']
dt=trainmodel(ohe_df,targetIndex,targetname,model)
dt.head(6)

DecisionTreeRegressor(criterion='mse', max_depth=20, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=3,
           min_samples_split=8, min_weight_fraction_leaf=0.0,
           presort=False, random_state=0, splitter='best')
val_mean squared error: 0.00
val_mean_absolute_error: 0.00
val_r2_score: 1.00
test_mean squared error: 29491498.45
test_mean_absolute_error: 162.35
test_r2_score: 0.96


predict     real
0   2398.329222   2398.0
1   6508.142189   6510.0
2   1260.008617   1260.0
3    894.435512    878.0
4   2994.245181   2995.0
5  27459.999781  27462.0

In [259]:
model = DecisionTreeRegressor(random_state=0,max_depth=20,min_samples_leaf=3,min_samples_split=8)
targetname=['sales_180days']
dt=trainmodel(ohe_df,targetIndex,targetname,model)
dt.head(6)

DecisionTreeRegressor(criterion='mse', max_depth=20, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=3,
           min_samples_split=8, min_weight_fraction_leaf=0.0,
           presort=False, random_state=0, splitter='best')
val_mean squared error: 0.00
val_mean_absolute_error: 0.00
val_r2_score: 1.00
test_mean squared error: 7996863.11
test_mean_absolute_error: 94.60
test_r2_score: 0.99


predict     real
0   2398.321794   2398.0
1   6510.555468   6510.0
2   1259.909053   1260.0
3   4343.073811   4390.0
4   2994.410918   2995.0
5  27462.000000  27462.0